vm/actor/src/builtin/miner/miner_state.rs

```diff
@@ -0,0 +1,993 @@
+    pub fn unlock_vested_funds<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        current_epoch: ChainEpoch,
+    ) -> Result<TokenAmount, Box<dyn StdError>> {
+        let mut vesting_funds = self.load_vesting_funds(store)?;
+        let amount_unlocked = vesting_funds.unlock_vested_funds(current_epoch);
+        self.locked_funds -= &amount_unlocked;
+        self.save_vesting_funds(store, &vesting_funds)?;
+        Ok(amount_unlocked)
+    }
+
+    /// CheckVestedFunds returns the amount of vested funds that have vested before the provided epoch.
+    pub fn check_vested_funds<BS: BlockStore>(
+        &self,
+        store: &BS,
+        current_epoch: ChainEpoch,
+    ) -> Result<TokenAmount, Box<dyn StdError>> {
+        let vesting_funds = self.load_vesting_funds(store)?;
+        Ok(vesting_funds
+            .funds
+            .iter()
+            .take_while(|fund| fund.epoch < current_epoch)
+            .fold(TokenAmount::zero(), |acc, fund| acc + &fund.amount))
+    }
+
+    /// Unclaimed funds that are not locked -- includes funds used to cover initial pledge requirement.
+    pub fn get_unlocked_balance(&self, actor_balance: &TokenAmount) -> TokenAmount {
+        let unlocked_balance = actor_balance - &self.locked_funds - &self.pre_commit_deposits;
+        assert!(unlocked_balance >= TokenAmount::zero());
+        unlocked_balance
+    }
+
+    /// Unclaimed funds. Actor balance - (locked funds, precommit deposit, ip requirement)
+    /// Can go negative if the miner is in IP debt.
+    pub fn get_available_balance(&self, actor_balance: &TokenAmount) -> TokenAmount {
+        // (actor_balance - &self.locked_funds) - &self.pre_commit_deposit
+        self.get_unlocked_balance(actor_balance) - &self.initial_pledge_requirement
+    }
+
+    pub fn assert_balance_invariants(&self, balance: &TokenAmount) {
+        assert!(self.pre_commit_deposits >= TokenAmount::zero());
+        assert!(self.locked_funds >= TokenAmount::zero());
+        assert!(*balance >= &self.pre_commit_deposits + &self.locked_funds);
+    }
+
+    pub fn meets_initial_pledge_condition(&self, balance: &TokenAmount) -> bool {
+        self.get_unlocked_balance(balance) >= self.initial_pledge_requirement
+    }
+
+    /// pre-commit expiry
+    pub fn quant_spec_every_deadline(&self) -> QuantSpec {
+        QuantSpec {
+            unit: WPOST_CHALLENGE_WINDOW,
+            offset: self.proving_period_start,
+        }
+    }
+
+    pub fn add_pre_commit_expiry<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        expire_epoch: ChainEpoch,
+        sector_number: SectorNumber,
+    ) -> Result<(), Box<dyn StdError>> {
+        // Load BitField Queue for sector expiry
+        let quant = self.quant_spec_every_deadline();
+        let mut queue = super::BitFieldQueue::new(store, &self.pre_committed_sectors_expiry, quant)
+            .map_err(|e| format!("failed to load pre-commit sector queue: {:?}", e))?;
```

Comment:
might want to downcast here if you feel it's important to add the message, otherwise might just be better to remove the map_err